arch/common/CMakeLists.txt

```diff
@@ -76,9 +76,9 @@ zephyr_linker_sources_ifdef(CONFIG_NOCACHE_MEMORY
   nocache.ld
 )
 
-# Only ARM, X86 and OPENISA_RV32M1_RISCV32 use ROM_START_OFFSET.
+# Only ARM, X86 and RISCV use ROM_START_OFFSET.
 if (DEFINED CONFIG_ARM OR DEFINED CONFIG_X86 OR DEFINED CONFIG_ARM64
-    OR DEFINED CONFIG_SOC_OPENISA_RV32M1)
+    OR (DEFINED CONFIG_RISCV AND CONFIG_ROM_START_OFFSET GREATER 0))
```

Comment:
why is `CONFIG_ROM_START_OFFSET > 0` needed here but only for risc?

Reply:
- I don't wanted to change it for the other archs
- there are some riscv soc, that don't work, so the part to add a offset is only used, if that offset is needed

Reply:
If the others are failing, they should be fixed, this is adding a hack

Reply:
the problem was mainly opentitan, which needs to put a other rom header in front. maybe @snematbakhsh can help

Reply:
@nordicjm the opentitan one is not compatible with having a CONFIG_ROM_START_OFFSET, because it has to put something soc specific in front in the linker.

Reply:
for the opentitan soc the offset is now forced to be `0x0`.

Reply:
> for the opentitan soc the offset is now forced to be 0x0.

so can the `CONFIG_ROM_START_OFFSET GREATER 0` be removed then ?
I agree with @nordicjm that this is not a nice adjustment.

Or can you elaborate so much on the issue that we might find another way to handle this ?

Reply:
no `CONFIG_ROM_START_OFFSET GREATER 0` is necessary , so that
```
  zephyr_linker_sources(ROM_START SORT_KEY ! rom_start_address.ld)
  # Some linkers fill unspecified region with pattern other than 0x00. Include
  # fill_with_zeros.ld file which forces the linker to use 0x00 pattern. Please
  # note that the pattern will affect empty spaces created after FILL(0x00).
  zephyr_linker_sources(ROM_START SORT_KEY $ fill_with_zeros.ld)
 zephyr_linker_sources(ROM_START SORT_KEY 0x0 rom_start_offset.ld)
```
is not used on the opentitan soc

Reply:
@maass-hamburg  Once we turn on CONFIG_ROM_START_OFFSET for risc-v, I think we need to account for the manifest (similar to [what is done here for another SOC](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/rw/Kconfig.defconfig#L8)). So I think we want to change [soc/lowrisc/opentitan/Kconfig.defconfig](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/lowrisc/opentitan/Kconfig.defconfig) as follows:

```
# The OpenTitan SoC requires a manifest in front of the
# application binary.
config ROM_START_OFFSET
	default 0x404
```
.. and subsequently we no longer need the "GREATER 0" special case.

Why 0x404 and not 0x400 for 1024 byte manifest, you ask? [It's this bug](https://github.com/zephyrproject-rtos/zephyr/issues/82822), I will take care of it once you land your PR (so just use 0x404).

Reply:
@snematbakhsh the `CONFIG_ROM_START_OFFSET` will result in everything before the offset being filled with zeros. But you need to write the start address in the manifest  see here: https://github.com/zephyrproject-rtos/zephyr/blob/90ff9c0883b29dc0c59e3dd063114e30a4b27456/soc/lowrisc/opentitan/rom_header.S#L25

We would also have to remove the part for creating the manifest, because it would conflict with the  `CONFIG_ROM_START_OFFSET` code.

Reply:
@maass-hamburg 

> the CONFIG_ROM_START_OFFSET will result in everything before the offset being filled with zeros.

It seems to work fine on my side, did I miss something?
Applying patches 1d65a59355fb53a606db1b8ffa54c73dc2800095 (to generate .bin) and 387478cdbdedb1de8aba526bb23298c5f7139156 (my proposed change as previously discussed). 

```
$ west build -b opentitan_earlgrey samples/hello_world/
$ hexdump build/zephyr/zephyr.bin | more
0000000 0000 0000 0000 0000 0000 0000 0000 0000
*
0000380 0000 0000 0404 0000 0000 0000 0000 0000
0000390 0000 0000 0000 0000 0000 0000 0000 0000
*
0000400 0000 0000 1197 f000 8193 4541 0297 0000
```

I think this works because [the SORT_KEY for the OpenTitan manifest](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/lowrisc/opentitan/CMakeLists.txt#L5) ("000") precedes [the SORT_KEY for rom_start_offset.ld](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/common/CMakeLists.txt#L89) ("0x0").

Reply:
seems like its working